File: app/src/main/java/com/example/dagger_hilt/domain/MovieRepository.java
package com.example.dagger_hilt.domain;

import androidx.lifecycle.LiveData;

import com.example.dagger_hilt.data.datasource.database.dao.MovieDao;
import com.example.dagger_hilt.data.datasource.database.entities.MovieEntity;
import com.example.dagger_hilt.data.datasource.web.api.MovieService;
import com.example.dagger_hilt.data.datasource.web.models.response.ApiResponse;
import com.example.dagger_hilt.data.datasource.web.models.MovieModel;
import com.example.dagger_hilt.data.datasource.web.models.MoviesResponse;
import com.example.dagger_hilt.sys.util.AppExecutors;
import com.example.dagger_hilt.sys.util.Constants;
import com.example.dagger_hilt.sys.util.Resource;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class MovieRepository {
    private final MovieDao dao;
    private final MovieService service;
    private final AppExecutors appExecutors;

    @Inject
    public MovieRepository(MovieDao dao, MovieService service, AppExecutors appExecutors) {
        this.dao = dao;
        this.service = service;
        this.appExecutors = appExecutors;
    }

    public LiveData<Resource<List<MovieEntity>>> loadMovies() {
        return new NetworkBoundResource<List<MovieEntity>, MoviesResponse>(appExecutors) {
            @Override
            protected void saveCallResult(MoviesResponse moviesResponse) {
                List<MovieEntity> moviesList = new ArrayList<>();
                for (MovieModel movie : moviesResponse.getListMovies()) {
                    moviesList.add(new MovieEntity(
                            movie.getId(),
                            movie.getTitle(),
                            movie.getOriginalTitle(),
                            movie.getOverview(),
                            movie.getPosterPath(),
                            false
                    ));
                }
                dao.setMovies(moviesList);
            }

            @Override
            protected boolean shouldFetch(List<MovieEntity> data) {
                return data == null || data.isEmpty();
            }

            @Override
            protected LiveData<List<MovieEntity>> loadFromDb() {
                return dao.getMovies();
            }

            @Override
            protected LiveData<ApiResponse> createCall() {
                return ApiResponse.create(service.loadMovies(Constants.API_KEY));
            }
        }.asLiveData();
    }

    public void updateMovie(int id, Boolean check) {
        dao.updateMovie(id, check);
    }
}

File: app/src/main/java/com/example/dagger_hilt/domain/MovieRepository.kt
package com.example.dagger_hilt.domain

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.example.dagger_hilt.data.datasource.database.dao.MovieDaoKts
import com.example.dagger_hilt.data.datasource.database.entities.MovieEntityKts
import com.example.dagger_hilt.data.datasource.web.api.MovieServiceKts
import com.example.dagger_hilt.data.datasource.web.models.ApiResponse
import com.example.dagger_hilt.data.datasource.web.models.MoviesResponseKts
import com.example.dagger_hilt.sys.util.AppExecutors
import com.example.dagger_hilt.sys.util.ConstantsKts
import com.example.dagger_hilt.sys.util.ResourceKts
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class MovieRepository @Inject constructor(
    private val dao: MovieDaoKts,
    private val service: MovieServiceKts,
    private val appExecutor: AppExecutors
) {
    fun loadMovies(): LiveData<ResourceKts<List<MovieEntityKts>>> =
        object : NetworkBoundResource<List<MovieEntityKts>, MoviesResponseKts>(appExecutor) {
            override fun saveCallResult(response: MoviesResponseKts) =
                response.listMovies.forEach { movie ->
                    dao.setMovie(
                        MovieEntityKts(
                            movie.id,
                            movie.title,
                            movie.originalTitle,
                            movie.overview,
                            movie.posterPath,
                            like = false
                        )
                    )
                }

            override fun shouldFetch(data: List<MovieEntityKts>?): Boolean = data.isNullOrEmpty()

            override fun loadFromDb(): LiveData<List<MovieEntityKts>> = dao.getMovies()

            override suspend fun createCall(): LiveData<ApiResponse<MoviesResponseKts>> =
                MutableLiveData(ApiResponse.create(service.loadMovies(ConstantsKts.API_KEY)))
        }.asLiveData()

    fun updateMovie(id: Int, check: Boolean) = dao.updateMovie(id, check)

}